sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -1065,32 +1088,38 @@ public boolean start() throws IOException {
       keyDeserializerInstance.configure(spec.getConsumerConfig(), true);
       valueDeserializerInstance.configure(spec.getConsumerConfig(), false);
 
-      for (PartitionState p : partitionStates) {
-        if (p.nextOffset != UNINITIALIZED_OFFSET) {
-          consumer.seek(p.topicPartition, p.nextOffset);
-        } else {
-          // nextOffset is unininitialized here, meaning start reading from latest record as of now
-          // ('latest' is the default, and is configurable) or 'look up offset by startReadTime.
-          // Remember the current position without waiting until the first record is read. This
-          // ensures checkpoint is accurate even if the reader is closed before reading any records.
-          Instant startReadTime = spec.getStartReadTime();
-          if (startReadTime != null) {
-            p.nextOffset =
-                consumerSpEL.offsetForTime(consumer, p.topicPartition, spec.getStartReadTime());
-            consumer.seek(p.topicPartition, p.nextOffset);
-          } else {
-            p.nextOffset = consumer.position(p.topicPartition);
+      // Seek to start offset for each partition. This is the first interaction with the server.
+      // Unfortunately it can block forever in case of network issues like incorrect ACLs.
+      // Initialize partition in a separate thread and cancel it if takes longer than a minute.
+      for (final PartitionState p : partitionStates) {
+        Future<?> future =  consumerPollThread.submit(new Runnable() {
```

Comment:
I think it is the other way round. Queuing them all at once is dependent on thread pool being single threaded. KafkaConsumer will throw an exception if two of those run in parallel, say when the pool is changed to allow more thread. Current code will work as long as there is one or more threads in the pool.

Cleaning up is also more involved when all of them are queued at once, we need to cancel all the remaining tasks. It is important to shutdown the thread pool properly. I forgot to mention earlier when you asked why we need to wakeup(). start() is called repeatedly forever on the worker in case of errors. If we don't shutdown the threadpool, we could keep accumulating threads. In this particular case, this pool in created in the Reader's constructor and shutdown in close(). If we don't wakeup() the stuck consumer, close() will block forever (it logs message every few seconds).